File: app_mobile/lib/core/models/shopping_list_item.dart
import 'product.dart';

class ShoppingListItem {
  final String id;
  final Product? product;
  final String? customName;
  final double? quantity;
  final String? unit;
  final bool purchased;

  ShoppingListItem({
    required this.id,
    this.product,
    this.customName,
    this.quantity,
    this.unit,
    required this.purchased,
  });

  String get displayName => product?.name ?? customName ?? '';

  factory ShoppingListItem.fromJson(Map<String, dynamic> json) => ShoppingListItem(
        id: json['id'] as String,
        product: json['product'] != null ? Product.fromJson(json['product'] as Map<String, dynamic>) : null,
        customName: json['customName'] as String?,
        quantity: (json['quantity'] as num?)?.toDouble(),
        unit: json['unit'] as String?,
        purchased: json['purchased'] as bool,
      );
}

File: backend/src/main/java/com/mycosplaymanager/backend/entity/ShoppingListItem.java
package com.mycosplaymanager.backend.entity;

import java.math.BigDecimal;
import java.time.Instant;
import java.util.UUID;

import org.hibernate.annotations.UuidGenerator;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.FetchType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.PrePersist;
import jakarta.persistence.Table;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

@Entity
@Table(name = "shopping_list_items")
@Getter
@Setter
@NoArgsConstructor
public class ShoppingListItem {

    @Id
    @UuidGenerator
    private UUID id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "team_id", nullable = false)
    private Team team;

    /** Nullable: si puo' segnare da comprare un prodotto non ancora nel catalogo (nome libero). */
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "product_id")
    private Product product;

    @Column(name = "custom_name")
    private String customName;

    private BigDecimal quantity;

    private String unit;

    @Column(name = "is_purchased", nullable = false)
    private boolean purchased = false;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "created_by", nullable = false)
    private User createdBy;

    @Column(name = "created_at", nullable = false, updatable = false)
    private Instant createdAt;

    @PrePersist
    void onCreate() {
        this.createdAt = Instant.now();
    }
}
